### PR TITLE
Settings GUI to edit Nammu's configuration

### PR DESCRIPTION
**Pull request by raquel-ucl**
Related to #206 and #217.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -36,18 +36,20 @@ of ORACC, enabling more projects to adopt it.
...
 ATF parser developed by UCL RSDT.
 
+Lemmatisation can only be done online with the ORACC server.
 
-#### Validation against the ORACC server
+
+#### Validation and lemmatisation against the ORACC server
 The ORACC server is hosted at University of Pennsylvania (Philadelphia) and
 maintained by [Prof. Steve Tinney](https://www.ling.upenn.edu/people/tinney).
 The ORACC server provides with SOAP web services for validation and lemmatisation
```
Comment (jonc125) [inline README.md]: Remove 'with' :)

Comment (jonc125) [inline README.md]: There's another extra 'with' in the following paragraph too - spotted as I was reading the rendered readme!

File: README.md [hunk 2]
```diff
@@ -77,11 +79,22 @@ on how to correct them.
 ## How to run Nammu
 
 To run Nammu on your computer, you can download it from
-[here](https://github.com/oracc/nammu/releases/download/0.5.0/nammu-0.5.5.jar)
+[here](https://github.com/oracc/nammu/releases/download/0.6/nammu-0.6.0.jar)
 and just double click on it to open it and use it.
 
-If you find any problem trying to open it, it might be that you don't have Java
-installed. You can check if you do by opening a terminal and typing:
+If you find any problem trying to open it,
```
Comment (jonc125) [inline README.md]: Looks like this line should be cut.

File: README.md [hunk 3]
```diff
@@ -96,32 +109,17 @@ If you keep having trouble, you can also run it from the command line, like this
...
-Please follow the guidelines to help us better understand and reproduce your
-problem.
-
-## Known problems and features under development
-We are currently working on the next release of Nammu, which includes some new
-features as well as bug fixes.
-
-Here's a list of known bugs:
+### Known problems:
 * Nammu doesn't validate or lemmatise properly when filenames include a space.
```
Comment (jonc125) [inline README.md]: Ouch! Do you know why this is?

File: README.md [hunk 4]
```diff
@@ -134,7 +132,13 @@ by setting a new environment variable called `NAMMU_CONFIG_PATH` that points to
...
-a YAML file containing some configuration on how the logging works.
+a [YAML file](https://github.com/oracc/nammu/blob/master/resources/config/logging.yaml)
+containing some configuration on how the logging works.
+
+There is also [a configuration file](https://github.com/oracc/nammu/blob/master/resources/config/settings.yaml)
+for shortcuts, preferred working directory, preferred list of ORACC projects,
+etc. This will be fully editable from Nammu, but for now only working directory
+and lemmatisation server can be edited from the settings window.
 
 You can see [here](https://github.com/oracc/nammu/blob/development/resources/config/logging.yaml)
```
Comment (jonc125) [inline README.md]: I'd move this sentence into the paragraph about the log file above.

File: python/nammu/SOAPClient/SOAPClient.py [hunk 5]
```diff
@@ -25,10 +25,11 @@
 import logging.config
 import requests
 import urllib
+import socket
```
Comment (jonc125) [inline python/nammu/SOAPClient/SOAPClient.py]: Is this used? I can't see it used in the diff.

File: python/nammu/SOAPClient/SOAPClient.py [hunk 6]
```diff
@@ -62,7 +63,12 @@ def send(self):
         self.logger.debug("Sending request to server at %s.", url)
         self.logger.debug("HTTP request headers sent: %s", headers)
         self.logger.debug("HTTP request body sent: %s", body)
-        self.response = requests.post(url, data=body, headers=headers)
+        try:
+            self.response = requests.post(url, data=body, headers=headers,
+                                          timeout=3)
+        except ConnectTimeout:
+            self.logger.error('Connetion timed out when sending POST request.')
```
Comment (jonc125) [inline python/nammu/SOAPClient/SOAPClient.py]: Spelling error - connection :)

File: python/nammu/controller/EditSettingsController.py [hunk 7]
```diff
@@ -0,0 +1,58 @@
+'''
+Copyright 2015, 2016 University College London.
```
Comment (jonc125) [inline python/nammu/controller/EditSettingsController.py]: It's 2017 now...

File: python/nammu/controller/EditSettingsController.py [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+        self.load_config()
+        self.view = EditSettingsView(self, self.working_dir, self.servers,
+                                     self.keystrokes, self.languages,
+                                     self.projects)
+        self.view.display()
+
+    def load_config(self):
+        '''
+        The user's config file should containg all necessary information for
+        the settings editor.
```
Comment (jonc125) [inline python/nammu/controller/EditSettingsController.py]: Or for Nammu as a whole? (Except for logging.)

File: python/nammu/controller/EditSettingsController.py [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+        '''
+        config_keywords = ['working_dir', 'servers', 'keystrokes',
+                           'languages', 'projects']
+        for keyword in config_keywords:
+            try:
+                setattr(self, keyword, self.config[keyword])
+            except KeyError:
+                self.controller.logger.error('%s missing on settings file.',
```
Comment (jonc125) [inline python/nammu/controller/EditSettingsController.py]: Should possibly document that all settings need to be in the config file. I agree that this approach is better than hiding defaults in the code though.

File: python/nammu/controller/EditSettingsController.py [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+                                             keyword)
+                self.view.display_error(keyword)
+
+    def update_config(self, working_dir, server, keystrokes=None,
+                      languages=None, projects=None):
+        '''
+        Update the settings file with the user input.
+        '''
+        # TODO: Validate new values introduced by user.
```
Comment (jonc125) [inline python/nammu/controller/EditSettingsController.py]: When you do this, the same validation should be applied to values read from the config file.

File: python/nammu/controller/EditSettingsController.py [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+        self.config['working_dir']['default'] = working_dir
+        self.config['servers']['default'] = server
+        self.controller.logger.debug("Settings updated.")
```
Comment (jonc125) [inline python/nammu/controller/EditSettingsController.py]: You haven't saved keystrokes, languages & projects.

File: python/nammu/controller/NammuController.py [hunk 12]
```diff
@@ -567,15 +571,20 @@ def send_request(self, client):
...
+                              'settings menu.')
+            raise Exception("ORACC server timed out after 5 seconds.")
+        except ConnectionError:
+            raise Exception("Can't connect to ORACC server at %s.",
+                            client.url)
         except HTTPError:
-            self.logger.error("ORACC server returned invalid HTTP response.")
-            raise
+            raise Exception("ORACC server returned invalid HTTP response.")
+        except ConnectTimeout:
```
Comment (jonc125) [inline python/nammu/controller/NammuController.py]: Merge with `Timeout` clause?

File: python/nammu/controller/NammuController.py [hunk 13]
```diff
@@ -567,15 +571,20 @@ def send_request(self, client):
...
-            self.logger.error("ORACC server timed out after 5 seconds.")
-            raise
-        except ConnectionError:
-            self.logger.error("Can't connect to ORACC server at %s.",
+            self.logger.error('ORACC %s server timed out after 5 seconds.',
                               client.url)
-            raise
+            self.logger.error('You can try with a different server from the '
+                              'settings menu.')
+            raise Exception("ORACC server timed out after 5 seconds.")
```
Comment (jonc125) [inline python/nammu/controller/NammuController.py]: A minor one, but might be worth considering your own class for errors raised explicitly by Nammu.

File: python/nammu/controller/NammuController.py [hunk 14]
```diff
@@ -648,7 +657,7 @@ def editSettings(self, event=None):
         settings form.
         TODO: Disable this button until functionality is implemented.
```
Comment (jonc125) [inline python/nammu/controller/NammuController.py]: Is the todo comment obsolete now?

File: python/nammu/view/EditSettingsView.py [hunk 15]
```diff
@@ -0,0 +1,269 @@
...
+            if server != "default":
+                combo_item = "{}: {}:{}".format(server,
+                                                self.servers[server]['url'],
+                                                self.servers[server]['port'])
+                combo.addItem(combo_item)
+                # If this item is the default one, set it as selected
+                if server == self.servers['default']:
+                    combo.setSelectedItem(combo_item)
+        # Make default server the selected item in combo box.
+        combo.setSelectedItem(self.servers['default'].split(':'))
```
Comment (jonc125) [inline python/nammu/view/EditSettingsView.py]: This is done in the previous line too!

File: python/nammu/view/EditSettingsView.py [hunk 16]
```diff
@@ -0,0 +1,269 @@
...
+            if server != "default":
+                combo_item = "{}: {}:{}".format(server,
+                                                self.servers[server]['url'],
+                                                self.servers[server]['port'])
+                self.combo.addItem(combo_item)
+                # If this item is the default one, set it as selected
+                if server == self.servers['default']:
+                    self.combo.setSelectedItem(combo_item)
+        # Make default server the selected item in combo box.
+        self.combo.setSelectedItem(self.servers['default'].split(':'))
```
Comment (jonc125) [inline python/nammu/view/EditSettingsView.py]: Repetition again?

File: python/nammu/view/EditSettingsView.py (hunk 15 above)
Comment (jonc125) [inline python/nammu/view/EditSettingsView.py]: I think this line doesn't work, actually, as `self.servers['default']` doesn't contain a colon.

File: python/nammu/view/EditSettingsView.py [hunk 17]
```diff
@@ -0,0 +1,269 @@
...
+        return combo
+
+    def build_servers_panel(self):
+        '''
+        Create panel that contains a drop down with the servers to choose from.
+        '''
+        panel = JPanel(FlowLayout())
+        label = JLabel("ORACC server location:")
+        panel.add(label)
+        self.combo = JComboBox()
```
Comment (jonc125) [inline python/nammu/view/EditSettingsView.py]: Shouldn't you use `build_servers_combobox` here?

**Comment by raquel-ucl** [comment]
Thanks a lot for this, @jonc125! 

Just a general sanity check is very good already. Have you tried to run it on your machine? You can checkout this branch, start a jython CLI and run it like this:

```python
from python.nammu.controller.NammuController import NammuController; nammu = NammuController()
```

I've just realised I should add a requirements file so you make sure you have everything you need to run it. Give me a minute...

**Comment by raquel-ucl** [comment]
Oh, and for the wireframes I used [Pencil](http://pencil.evolus.vn), same one I used for CloudLabs.

File: README.md (hunk 3 above)
Comment (raquel-ucl) [inline README.md]: Yep. That's the ORACC server not liking files with spaces on their filenames. We'll be working on the backend later on in the project possible and could address this.

File: README.md (hunk 2 above)
Comment (raquel-ucl) [inline README.md]: Well spotted, I forgot to add "have a look at the Troubleshooting section" and link to it.

File: python/nammu/SOAPClient/SOAPClient.py (hunk 5 above)
Comment (raquel-ucl) [inline python/nammu/SOAPClient/SOAPClient.py]: Nop. Leftovers from testing different things. Atom should have complained!

File: python/nammu/controller/EditSettingsController.py (hunk 8 above)
Comment (raquel-ucl) [inline python/nammu/controller/EditSettingsController.py]: The configuration in the config file is for Nammu as a whole, but all the configuration bits needed by this specific class are in the config file. Does this make sense? I'll try to rephrase.

File: python/nammu/controller/EditSettingsController.py (hunk 9 above)
Comment (raquel-ucl) [inline python/nammu/controller/EditSettingsController.py]: I do need to write a better user manual for ORACC. I've created #218 and should document better the contents of the config files there.

File: python/nammu/controller/EditSettingsController.py (hunk 11 above)
Comment (raquel-ucl) [inline python/nammu/controller/EditSettingsController.py]: Those are not implemented yet. I've added a TODO message for that.

File: python/nammu/controller/EditSettingsController.py (hunk 10 above)
Comment (raquel-ucl) [inline python/nammu/controller/EditSettingsController.py]: Yes. For now I'm only checking that the yaml file is well built. Then each field is validated when they are used by different classes. Probably better to have a validation method that is called when the yaml is loaded. I'll create an issue about this.

File: python/nammu/controller/NammuController.py (hunk 14 above)
Comment (raquel-ucl) [inline python/nammu/controller/NammuController.py]: Well spotted!

**Comment by jonc125** [comment]
I tried creating an anaconda env for Nammu, but can't install Jython in such an env, and the system jython uses the system site-packages so doesn't find pyyaml.

Incidentally, the version constraints in the requirements.txt don't match those used by `mvn` (<= rather than <):
```
[INFO] installing requested python libraries
[INFO] running [java, -cp, .:Lib, -Dpython.home=., -Dpython.executable=jython, org.python.util.jython, Lib/site-packages/setuptools-0.6c11-py2.7.egg/easy_install.py, --build-directory, /Users/jonc/projects/nammu/target/jython-plugins-tmp/build, requests<2.12.0, pytest<3.0, pyyaml, mako, ply, https://github.com/ucl/pyoracc/archive/master.tar.gz] in target/jython-plugins-tmp
```

**Comment by jonc125** [comment]
Some thoughts from running the jar built by `mvn`:
* Took me some hunting to find the settings menu item - 'help' was the last place I checked!
* Seeing 'Working directory: None' surprised me. If this defaults to the cwd, I'd expect to see that path shown.

**Comment by raquel-ucl** [comment]
To create my virtual environment, I used [virtualenvwrapper](http://virtualenvwrapper.readthedocs.io/en/latest/index.html) and passed the system's jython (which I installed with brew) like this:

```bash
mkvirtualenv -p python nammu
```
Not sure if that'll mess up your python environment much, because I don't use conda :-/

**Comment by raquel-ucl** [comment]
* I'll make working_dir default somewhere other than None. I had changed mine manually in the yaml file, so I didn't realise about the default!

* I'll move the settings to the edit menu. That makes more sense I think.

File: python/nammu/controller/EditSettingsController.py (hunk 7 above)
Comment (raquel-ucl) [inline python/nammu/controller/EditSettingsController.py]: `sed` to the rescue. Copyright updated.

File: python/nammu/view/EditSettingsView.py (hunk 17 above)
Comment (raquel-ucl) [inline python/nammu/view/EditSettingsView.py]: This has helped me realise there were two left over methods from when I decided to change the layout.

**Comment by jonc125** [comment]
Using virtualenv Jython is still not finding packages, alas. If I do `sys.path.append('/Users/jonc/projects/nammu/venv/lib/python2.7/site-packages')` I then get
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "python/nammu/controller/NammuController.py", line 29, in <module>
    from AtfAreaController import AtfAreaController
  File "python/nammu/controller/AtfAreaController.py", line 27, in <module>
    import TextLineNumber
ImportError: No module named TextLineNumber
```
so l think there's some other path setup missing somewhere.

Still, I can run the jar built by maven so that's not a serious issue. For later perhaps worth writing a contributing.md with more info on how to get set up to contribute (and extract those bits that already exist from the readme).

**Comment by raquel-ucl** [comment]
Ah, that's a different problem. There is a Java class file that you need to be able to access. It's in `./target/classes/resources/lib/`, so if you add that folder to your class path, I think that should fix the problem.

File: python/nammu/controller/EditSettingsController.py (hunk 10 above)
Comment (raquel-ucl) [inline python/nammu/controller/EditSettingsController.py]: I'm just created #220

**Comment by raquel-ucl** [comment]
@jonc125 I think it's all done now. Things that are not addressed are in new issues to be addressed later. Thanks so much for your help!

**Comment by EleanorRobson** [comment]
I'd just like to chip in to say THANK YOU to both of you! I've been following this out of the corner of my eye all day, and am hugely grateful that you've been so responsive. Hope you both have a great weekend!

**Pull request merged**